src/mq/gcloud-pubsub-push/mq.ts

```diff
@@ -0,0 +1,362 @@
+    private errorListener?: (error: Error) => void;
+
+    /**
+     * Creates a new message queue
+     *
+     * @param logger {Logger} Logger instance
+     * @param pubSubClient {PubSub} Pub/Sub client instance
+     * @param topic {string} Full name of topic to publish messages to
+     */
+    constructor(logger: Logger, pubSubClient: PubSub, topic: string) {
+        this.logger = logger;
+        this.pubSubClient = pubSubClient;
+        this.topic = topic;
+    }
+
+    /**
+     * Indicates whether the message queue is listening for messages or not
+     */
+    get isListening(): boolean {
+        return this.messageHandler !== undefined;
+    }
+
+    /**
+     * Enqueues a message
+     *
+     * @param message {any} Message to enqueue
+     * @param options {MessageQueueEnqueueOptions} Options for the enqueue operation
+     */
+    async enqueue(
+        message: any,
+        options?: MessageQueueEnqueueOptions,
+    ): Promise<void> {
+        const delay = options?.delay?.total('millisecond');
+
+        // If the message has a delay, do not enqueue it - This is likely a retry
+        // attempt by Fedify, but we do not want to retry the message as we want
+        // to use GCloud Pub/Sub's built in retry mechanism
+        if (delay !== undefined) {
+            this.logger.info(
+                `Not enqueuing message [FedifyID: ${message.id}] due to delay being set: ${delay}`,
+                { fedifyId: message.id },
+            );
+
+            return;
+        }
+
+        this.logger.info(`Enqueuing message [FedifyID: ${message.id}]`, {
+            fedifyId: message.id,
+        });
+
+        try {
+            const messageId = await this.pubSubClient
+                .topic(this.topic)
+                .publishMessage({
+                    json: message,
+                    attributes: {
+                        fedifyId: message.id,
+                    },
+                });
+
+            this.logger.info(
+                `Message [FedifyID: ${message.id}] was enqueued [PubSubID: ${messageId}]`,
+                { fedifyId: message.id, pubSubId: messageId },
+            );
+        } catch (error) {
+            this.logger.error(
+                `Failed to enqueue message [FedifyID: ${message.id}]: ${error}`,
+                { fedifyId: message.id, error },
+            );
+
+            this.errorListener?.(error as Error);
+        }
+    }
+
+    /**
+     * Starts the message queue
+     *
+     * @param handler {function} Message handler
+     * @param options {MessageQueueListenOptions} Options for the listen operation
+     */
+    async listen(
+        handler: (message: any) => Promise<void> | void,
+        options: MessageQueueListenOptions = {},
+    ): Promise<void> {
+        this.messageHandler = handler;
+
+        return await new Promise((resolve) => {
+            options.signal?.addEventListener('abort', () => {
+                resolve();
```

Comment:
I think we need to unassign `this.messageHandler` here? Or more specifically have some way of making `this.isListening` evaluate to false

Reply:
👍 Updated in https://github.com/TryGhost/ActivityPub/pull/180/commits/12ea3f6c2a592156b7a771b7be3c629b09612dbe